### PR TITLE
code review [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by doudou**

### CODE REVIEW

File: gazebo.orogen [hunk 1]
```diff
@@ -1,64 +0,0 @@
...
-##    output_port "output", "int"
-
-#    # If you want that component's updateHook() to be executed when the "input"
-#    # port gets data, uncomment this and comment the 'periodic' line
-#    # port_driven "input"
-#    # By default, the task will be periodic with a period of 0.1
-#    periodic 0.1
-#end
-
-task_context "ModelTask" do
```
Comment (doudou) [inline gazebo.orogen:62]: Use the template above to create this. In particular the first line (where it says that needs_configuration should be left there)

File: tasks/ModelTask.cpp [hunk 2]
```diff
@@ -1,198 +0,0 @@
...
-	
-//	Export Gazebo Joints and Links to Rock-Robotics
-	setJoints();
-	setLinks();
-} 
-//======================================================================================
-void gazebo::ModelTask::setJoints()
-{
-	typedef gazebo::physics::Joint_V Joint_V;
-	Joint_V joints;
```
Comment (doudou) [inline tasks/ModelTask.cpp:51]: Declare-then-use is C, in C++ one declares variables where they are used

```
Joint_V joints = model->GetJoints();
```

File: tasks/ModelTask.cpp [hunk 3]
```diff
@@ -1,198 +0,0 @@
...
-	{
-		gzmsg <<"RockBridge: found joint: "<< world->GetName() + "/" + model->GetName() + 
-				"/" + (*joint)->GetName() << std::endl;
-		gzmsg << "RockBridge: create joint InputPort in Rock." << std::endl;
-		joint_port = new RTT::InputPort<double>( "joint: " + world->GetName() +
-				"/" + model->GetName() + "/" + (*joint)->GetName() );
-		ports()->addPort( *joint_port );
-		joint_port_list.push_back( std::make_pair(joint_port,*joint) );			
-	}
-	joints.clear();
```
Comment (doudou) [inline tasks/ModelTask.cpp:65]: Why ?

File: tasks/ModelTask.cpp [hunk 4]
```diff
@@ -1,198 +0,0 @@
...
-} 
-//======================================================================================
-void gazebo::ModelTask::setJoints()
-{
-	typedef gazebo::physics::Joint_V Joint_V;
-	Joint_V joints;
-			
-	// Get all joints from a model and creates a Rock component InputPort
-	joints = model->GetJoints();
-	for(Joint_V::iterator joint = joints.begin(); joint != joints.end(); ++joint)
```
Comment (doudou) [inline tasks/ModelTask.cpp:55]: Sets of joints in Rock are nowadays gathered in a single base::samples::Joints structure. Just do that. You can then avoid creating dynamic ports and instead declare the 'joints' (output) and "cmd_joints" (input) ports in the orogen file

File: tasks/ModelTask.cpp [hunk 5]
```diff
@@ -1,198 +0,0 @@
...
-	{
-		gzmsg <<"RockBridge: found joint: "<< world->GetName() + "/" + model->GetName() + 
-				"/" + (*joint)->GetName() << std::endl;
-		gzmsg << "RockBridge: create joint InputPort in Rock." << std::endl;
-		joint_port = new RTT::InputPort<double>( "joint: " + world->GetName() +
```
Comment (doudou) [inline tasks/ModelTask.cpp:60]: Why did you add the world and model names (and "link") to the port name ? There is one component per model, a.k.a. no risk of name collision. And the link: thing breaks naming conventions (which is using alphanumeric characters and _ in port/property/operation names)

File: tasks/ModelTask.cpp [hunk 6]
```diff
@@ -1,198 +0,0 @@
...
-				"/" + model->GetName() + "/" + (*joint)->GetName() );
-		ports()->addPort( *joint_port );
-		joint_port_list.push_back( std::make_pair(joint_port,*joint) );			
-	}
-	joints.clear();
-}
-//======================================================================================
-void gazebo::ModelTask::setLinks()
```
Comment (doudou) [inline tasks/ModelTask.cpp:68]: Mostly same comments than with setJoints

File: tasks/ModelTask.cpp [hunk 7]
```diff
@@ -1,198 +0,0 @@
...
-	model = _model;
-	environment = _environment;
-//	sdf = model->GetSDF();
-	
-//	Export Gazebo Joints and Links to Rock-Robotics
-	setJoints();
-	setLinks();
-} 
-//======================================================================================
-void gazebo::ModelTask::setJoints()
```
Comment (doudou) [inline tasks/ModelTask.cpp:48]: setJoints is a weird name given what the method does.

File: tasks/ModelTask.cpp [hunk 8]
```diff
@@ -1,198 +0,0 @@
...
-void gazebo::ModelTask::updateLinks()
-{
-	// Update all links from gazebo model
-	for(LinkPort_V::iterator it = link_port_list.begin(); it != link_port_list.end(); ++it)
-	{
-		physics::LinkPtr link = (*it).second;
-
-		// Apply bouyancy to links in underwater environment
-		if(environment == UNDERWATER){
-			math::Vector3 cobPosition;
```
Comment (doudou) [inline tasks/ModelTask.cpp:127]: As I said during the meeting, this should really go in gazebo directly. Are there any ways to extend / update these parameters there ? Maybe have a plugin for fluid interaction ?

File: tasks/ModelTask.cpp (hunk 5 above)
Comment (doudou) [inline tasks/ModelTask.cpp:60]: This is a red flag. You assign to a pointer that is not a local variable within a loop. Which means that (1) either you have a memory leak or (2) you have a potential problem with double-free. In your case, you get the double-free.

Define your variables in the most narrow scope possible: where you start using them. Whenever you read your code, always ask yourself when you assign to "permanent" variables (i.e. object attributes)

File: tasks/ModelTask.cpp [hunk 9]
```diff
@@ -1,198 +0,0 @@
...
-			
-	}
-	links.clear();
-}
-//======================================================================================
-void gazebo::ModelTask::updateModel()
-{
-	// gzmsg << " ModelTask::updateModel() from model:"<< model->GetName() << std::endl;
-	
-	updateJoints();
```
Comment (doudou) [inline tasks/ModelTask.cpp:94]: Good reflex to split the updateModel method in two smaller disjoint parts

File: tasks/ModelTask.cpp [hunk 10]
```diff
@@ -1,198 +0,0 @@
...
-		(*it).first->readNewest(effort);
-		if(effort != RTT::NoData)
-		{
-			// Define the limits for the joint effort
-			if((-1.0 <= effort) && (effort <= 1.0)) 
-			{	
-				(*it).second->SetForce(0,effort);
-			} else
-			{
-				gzmsg << "RockBridge: error - effort value out of range (-1 <= effort <= 1)." << std::endl;
```
Comment (doudou) [inline tasks/ModelTask.cpp:112]: This should cause the component to go into exception. It should not happen, this is a fault of the component, do not allow it to go unnoticed.

File: tasks/ModelTask.cpp (hunk 8 above)
Comment (doudou) [inline tasks/ModelTask.cpp:127]: Moreover, it looks weird. The only link that you should update is the world-to-body link. You update all links with viscosity and such, which makes no sense to me. Any insights ?

File: tasks/ModelTask.cpp [hunk 11]
```diff
@@ -1,198 +0,0 @@
...
-			
-	// Get all links from a model and creates a Rock component InputPort
-	links = model->GetLinks();
-	for(Link_V::iterator link = links.begin(); link != links.end(); ++link)
-	{
-		gzmsg << "RockBridge: found link: "<< world->GetName() + "/" + model->GetName() + 
-				"/" + (*link)->GetName() << std::endl;
-		
-		gzmsg << "RockBridge: create link InputPort in Rock." << std::endl;
-		link_port = new RTT::InputPort<base::Vector3d>( "link: " + world->GetName() +
```
Comment (doudou) [inline tasks/ModelTask.cpp:81]: You should create a base::Force data structure in base/types for this.

File: tasks/ModelTask.cpp [hunk 12]
```diff
@@ -1,198 +0,0 @@
...
-				"/" + model->GetName() + "/" + (*link)->GetName() );
-		ports()->addPort(*link_port);
-		link_port_list.push_back( std::make_pair(link_port,*link) );
```
Comment (doudou) [inline tasks/ModelTask.cpp:84]: And where are the output ports ? Here, probably a RigidBodyState. For the links, also a base::samples::Joints

File: tasks/ModelTask.cpp [hunk 13]
```diff
@@ -1,198 +0,0 @@
...
-		
-			link->AddForceAtWorldPosition(link_bouyancy, cobPosition);											
-//			link->AddForceAtWorldPosition(resultant_force, cobPosition);
-//			link->AddForce(link_bouyancy);
-		}
-
-		// Read rock input port and apply a force to gazebo links
-		base::Vector3d force(0.0, 0.0, 0.0);
-		(*it).first->readNewest(force);				
-		if((force(0) != RTT::NoData)||(force(1) != RTT::NoData)||(force(2) != RTT::NoData)){
```
Comment (doudou) [inline tasks/ModelTask.cpp:193]: This is absolutely not how it works. Look at the tutorials again.

File: tasks/ModelTask.cpp (hunk 12 above)
Comment (Thomio-Watanabe) [inline tasks/ModelTask.cpp:84]: Do you mean I should use a gazebo method to read position of links and write them in a Rock output port !?

File: tasks/ModelTask.cpp (hunk 10 above)
Comment (Thomio-Watanabe) [inline tasks/ModelTask.cpp:112]: The user feed the input port and he doesn't know the input port acceptable range... how is this a fault of the component ?!
What should I do exactly ?

File: tasks/ModelTask.cpp (hunk 12 above)
Comment (doudou) [inline tasks/ModelTask.cpp:84]: Yes. Position of links and joints. Links being the transformation between the link joints (as base::samples::RigidBodyState), joints as base::samples::Joints.

The former is super useful if you want to compare e.g. "ground truth" with estimated values. The latter will basically be used as the joints sensor

File: tasks/ModelTask.cpp (hunk 10 above)
Comment (doudou) [inline tasks/ModelTask.cpp:112]: > The user feed the input port and he doesn't know the input port acceptable range... how is this a fault of the component ?!

You have a very good point, actually. It is _not_ a fault of this component. However, it is the generally accepted behaviour in rock to go into exception if the input is unacceptable.

**Comment by doudou** [comment]
This is pretty irrelevant now, I close it
